Create TI-Gives+Gets.md (#226)
ossf/tac: process/TI-Gives+Gets.md

# "Gives and Gets" for OpenSSF Technical Initiatives (TI)

> The OpenSSF has a large community of contributors and efforts that span the broad spectrum of open source security interests.  The Technical Initiatives (TIs) of the foundation are where our members collaborate and help craft unique solutions to addressing improving the security of the open source ecosystem.

Reviewer: Do we want to also highlight that his gives and gets is a bit vague on details and how someone should approach understanding more?

> May post project updates and tutorials to the OpenSSF blog.

How many? How often?

I know we don't have an answer to that but we might want to highlight that this is access to these benefits but the details will depend on time, scope, staffing, etc.

Author: I would suggest to make the guidance as broad as possible.  I'm not sure telling someone they can post 4 blogs a year really adds anything.  If folks have work output to share that adds value, they should get it shared; if they don't or on working on a long-running project, that's ok too.  But that's just my opinion.

> In exchange for meeting certain requirements, the TIs are eligible to receive an assortment of benefits and have access to the capabilities of the Foundation's resources.  The specific requirements and benefits (aka "Gives and Gets") for each level of maturity are documented below.  Based on the specific type of work the TI is focused on (e.g a software project vs. a specification or documentation-based effort) the requirements and benefits may slightly differ as applicable.

Reviewer 2: Done.

## Sandbox level Gives & Gets

> |     Gives/Requirements             |                Gets/Benefits          |

Reviewer: Nit: The table layout is a bit confusing because in the second table it doesn't line up similarly to the other two tables, e.g. just two lists. It makes it looks like the the individual gives and gets are potentially related to each other.

Author: Patches and formatting assistance welcome

Reviewer 2: Fixed

| :-----------------------------:    | :-----------------------------------: | 
|  TI must be aligned with the OpenSSF mission and either be a novel approach for existing areas or address an unfulfilled need. It is expected that the initial code needed for an OpenSSF WG to work be kept within their repository and will not function as a project in its own right. Should initial WG code grow and mature that it warrants its own Project status, then it is subject to Sandbox entry requirements. It is preferred that extensions of existing OpenSSF projects collaborate with the existing project rather than seek a new project. |    TI can get assistance with Architecture & Roadmap Alignment      | 
|  TI must maintain a diversified contributor base (i.e. not a single-vendor project).  TI must have a minimum of two maintainers with different organization affiliations. | Receives consideration as in-scope for any submission to an OpenSSF-managed conference or event. |
|  TI must find an aligned WG to host the TI and must have a TAC sponsor that can help guide the TI through processes.  | TI receives guidance on technical direction from TAC sponsor  |

> |  TI agrees to follow the [Secure Software Development Guiding Principles](https://github.com/ossf/wg-best-practices-os-developers/blob/main/docs/SecureSoftwareGuidingPrinciples.md) and the [Open Source Consumption Manifesto](https://github.com/ossf/wg-endusers/tree/main/MANIFESTO). | Receives OpenSSF Code of Conduct Committee support.|

Reviewer 2: Removed

|  If contributing an existing Project to the OpenSSF, the contribution must undergo license and IP due diligence by the Linux Foundation (LF). | Reserved space for project updates in OpenSSF newsletters.|
|  Provides quarterly updates to the TAC on technical vision and progress on vision.  | May request infrastructure support from the OpenSSF. |

> |  TI will have a [SECURITY.md](http://security.md/) that describes how the Project manages vulns, or more broadly how the OSSF handles vuln reports  |  Projects may say they are, "A sandbox project in the OpenSSF" or "An experimental project in the OpenSSF."  Gets an "sandbox" logo that is shared amongst all OpenSSF sandbox TIs.  |

Reviewer: > Gets an "sandbox"

a "sandbox"

Reviewer 2: Fixed

|  Projects must have documented project governance and be able to demonstrate that governance in action.  |  May request OpenSSF budget for project improvements such as security audits or time-bound contracting needs.  |
|  When applicable, projects must have completed a security audit through a third party and addressed audit findings and recommendations.  |  May request OpenSSF budget for sustained maintainer stipends (details determined by OpenSSF and project leads).  |
|   When applicable, Projects should achieve **BLAH** level of SLSA  | With additional TAC or WG approval, may fundraise for dedicated project funds, coordinated by the OpenSSF.|
|  |  Projects may use the OpenSSF logo to promote their project (in accordance with the trademark guidelines). Projects may be referred to as an "OpenSSF Project" or "OpenSSF $ProjectName."  |
|  |  May request considered for Grants  |  
|  |  May request consideration to get Contract Developers  |

> |  |  Requests for one time funding needs to include: Tech writer, Graphic designer, Security audit, Event support, Outreach, Dashboard/reports, Recognition awards, Infrastructure support for software projects (like special build, QA) (will have ongoing operational expense),  Cloud hosting (will have ongoing operational expense)  |

Reviewer: A lot of this feels like it should be under incubating.  If the goal of graduation from OpenSSF end is to have a well structured, well governed, well written, well documented project some of these projects are coming to OpenSSF to support there. I think things like security audits and documentation in particular should be things that OpenSSF helps provide to take a project from incubating to graduation.

Author: sounds fine to me, do you want to suggest how/where to move the text?

# "Gives and Gets" for OpenSSF Technical Initiatives (TI)

The OpenSSF has a large community of contributors and efforts that span the broad spectrum of open source security interests.  The Technical Initiatives (TIs) of the foundation are where community members collaborate and help craft unique solutions to address improving the security of the open source ecosystem.

> In exchange for meeting certain requirements, the TIs are eligible to receive an assortment of benefits and have access to the capabilities of the Foundation's resources.  The specific requirements and benefits (aka "Gives and Gets") for each level of maturity are documented below.

Reviewer 2: ```suggestion
In exchange for meeting certain requirements, the TIs are eligible to receive an assortment of benefits and have access to the capabilities of the Foundation's resources.  The specific requirements and benefits (aka "Gives and Gets") for each level of maturity are documented below. Note that this about the maturity of the TI itself - basically about how it operates - rather than of its product. A TI can be at an advanced level of maturity even though its product is still immature and vice versa.
```

 * TI must find an aligned WG to host the TI and must have a TAC sponsor that can help guide the TI through processes.
 * TI agrees to follow the [Secure Software Development Guiding Principles](https://github.com/ossf/wg-best-practices-os-developers/blob/main/docs/SecureSoftwareGuidingPrinciples.md) and the [Open Source Consumption Manifesto](https://github.com/ossf/wg-endusers/tree/main/MANIFESTO).
 * If contributing an existing Project to the OpenSSF, the contribution must undergo license and IP due diligence by the Linux Foundation (LF).
 * Provides quarterly updates to the TAC on technical vision and progress on vision.
 * TI will have a [SECURITY.md](https://github.com/ossf/project-template/blob/main/SECURITY.md) that describes how the Project manages vulnerabilities, or more broadly how the OSSF handles vulnerability reports

### Gets/Benefits

 * TI can get assistance with Architecture & Roadmap Alignment
 * Receives consideration as in-scope for any submission to an OpenSSF-managed conference or event.
 * TI receives guidance on technical direction from TAC sponsor
 * Reserved space for project updates in OpenSSF newsletters.

>  * May request infrastructure support from the OpenSSF.

Reviewer 2: ```suggestion
 * May request basic infrastructure support from the OpenSSF (e.g., mailing list and github repo).
```

 * Projects must have defined a contributor guide, which makes it clear how and when contributors should be given increasing responsibilities towards maintainership of the project. (Example guides: Sigstore, AllStar)
 * Projects should be able to show adoption by multiple parties and adoption's value to the open source community and/or end users (may include adoption of beta/early versions) with the intent to showcase wide adoption by the project's consumers.
 * TI must have documented, initial group governance.
 * Maintains a point of contact for vulnerability reports in the security.md
 * Implements, practices, and refines mature software development and release practices such as following a version schema.
 * TI follows security best practices (as recommended by the OpenSSF and others), including passing the OpenSSF Best Practices criteria, secret scanning, and code scanning.
 * TIs that include code use Scorecards
 * Begins to establish the appropriate governance that enables its sustainment for potential graduation.

### Gets/Benefits

 TI eligible to receive all Gets from Sandbox plus:

>  * Receives infrastructure support

Reviewer 2: ```suggestion
 * Receives more infrastructure support from the OpenSSF (e.g., website support)
```

 * May post project updates and tutorials to the OpenSSF blog.

## Graduated level Gives & Gets

### Gives/Requirements

 All requirements of Incubating must be fulfilled and additionally:
 * Projects must be able to show a consistent release cadence.
 * Maintains a point of contact for vulnerability reports and follow coordinated vulnerability disclosure practices.
 * Implements, practices, and refines mature software development and release practices, such as adherence to semantic versioning, and having a declared policy for stable releases and backported fixes.
 * Projects must have documented project governance and be able to demonstrate that governance in action.
 * When applicable, projects must have completed a security audit through a third party and addressed audit findings and recommendations.
 * Projects should harden their build systems in accordance with the SLSA Framework

### Gets/Benefits

 TI eligible to receive all Gets from Incubating plus:
 * Receives consideration as in-scope for any submission to an OpenSSF-managed conference or event.  This includes consideration for booth space at the conference and/or the OpenSSF booth.

>  * Receives infrastructure support (details determined by project leads and OpenSSF Budget Committee).

Reviewer 2: ```suggestion
 * Receives advanced infrastructure support from the OpenSSF (e.g., cloud hosting, to be determined by project leads and OpenSSF Budget Committee).
```

Based on the specific type of work the TI is focused on (e.g., software, specification, or documentation development) the requirements and benefits may slightly differ as applicable.

Also note that benefits may actually vary based on resources and funds availability, or lack thereof.

## Sandbox level Gives & Gets

### Gives/Requirements

 * TI must be aligned with the OpenSSF mission and either be a novel approach for existing areas or address an unfulfilled need. It is expected that the initial code needed for an OpenSSF WG to work be kept within their repository and will not function as a project in its own right. Should initial WG code grow and mature that it warrants its own Project status, then it is subject to Sandbox entry requirements. It is preferred that extensions of existing OpenSSF projects collaborate with the existing project rather than seek a new project.
 * TI must maintain a diversified contributor base (i.e. not a single-vendor project).  TI must have a minimum of two maintainers with different organization affiliations.
 * TI must find an aligned WG to host the TI and must have a TAC sponsor that can help guide the TI through processes.

>  * TI agrees to follow the [Secure Software Development Guiding Principles](https://github.com/ossf/wg-best-practices-os-developers/blob/main/docs/SecureSoftwareGuidingPrinciples.md) and the [Open Source Consumption Manifesto](https://github.com/ossf/wg-endusers/tree/main/MANIFESTO).

Reviewer 3: Is it worth tagging these with versions (or referencing a commit SHA instead of pointing to main)? I'm not sure how often these will change but, making this a "Give" should put some burden on the groups producing these to solicit feedback of proposed future changes with the TIs.

 * If contributing an existing Project to the OpenSSF, the contribution must undergo license and IP due diligence by the Linux Foundation (LF).
 * Provides quarterly updates to the TAC on technical vision and progress on vision.

>  * TI will have a [SECURITY.md](http://security.md/) that describes how the Project manages vulnerabilities, or more broadly how the OSSF handles vulnerability reports

Reviewer 4: ```suggestion
 * TI will have a [SECURITY.md](https://github.com/ossf/project-template/blob/main/SECURITY.md) that describes how the Project manages vulnerabilities, or more broadly how the OSSF handles vulnerability reports
```

> ### Gets/Benefits

Reviewer 5: Awesome :). Looking forward to more information on how to request assistance!

> ### Gets/Benefits

Reviewer 4: There's definitely some overlap here in items and wording from the other files in `process/` (e.g. https://github.com/ossf/tac/blob/main/process/project-lifecycle.md). Are we okay with that duplication? Or do we want to focus on things that aren't already covered in lifecycle process documentation?

Reviewer 2: Indeed. This document was started with what's in the project-lifecycle.md. The plan (that I suspected and CRob has now confirmed to me) is that this is meant to be a generalized version that will be applicable to all types of TIs. Once this is completed we'll remove the overlapping pieces from the project-lifecycle.md file and point to this file instead.

 * TI can get assistance with Architecture & Roadmap Alignment
 * Receives consideration as in-scope for any submission to an OpenSSF-managed conference or event.
 * TI receives guidance on technical direction from TAC sponsor

>  * Reserved space for project updates in OpenSSF newsletters.

Reviewer 6: Could we also update the OpenSSF website with a project page similar to CNCF to promote the projects? I have created an issue for this: https://github.com/ossf/tac/issues/222

### Gives/Requirements

 All requirements of Sandbox must be fulfilled. PR filed to promote group to Incubating stage.
 * Group has met no less than 5 times within the last calendar quarter
 * Maintains a diversified contributor base (i.e. not a single-vendor project) with an active flow of contributions.  Projects must have a minimum of three maintainers with a minimum of two different organization affiliations, and document the current list of maintainers.
 * Projects must have defined a contributor guide, which makes it clear how and when contributors should be given increasing responsibilities towards maintainership of the project. (Example guides: Sigstore, AllStar)
 * Projects should be able to show adoption by multiple parties and adoption's value to the open source community and/or end users (may include adoption of beta/early versions) with the intent to showcase wide adoption by the project's consumers.
 * TI must have documented, initial group governance.
 * Maintains a point of contact for vulnerability reports in the security.md
 * Implements, practices, and refines mature software development and release practices such as following a version schema.

>  * TI Follows security best practices (as recommended by the OpenSSF and others), including passing the OpenSSF Best Practices criteria

Reviewer 4: ```suggestion
 * TI follows security best practices (as recommended by the OpenSSF and others), including passing the OpenSSF Best Practices criteria, secret scanning, and code scanning.
```

>  * Project should be integrating with Scorecards

Reviewer 4: Does "should" mean that these requirements are recommended but optional?

>  * Project should be integrating with Scorecards

Reviewer 4: ```suggestion
 * TIs that include code use Scorecards
```
To disambiguate if the "should" means "only relevant to TIs that include code" or "it's recommended, but not required, for TIs with code to use Scorecards".

 * Begins to establish the appropriate governance that enables its sustainment for potential graduation.

>  * Projects should be Securing Code Repository -> Managing Contributions Commit Signing , Secret Scanning, Code Scanning (OSFUZZ at a minimum) + Self-assessment Should OpenSSF require these if the SCM supports it, especially using Sigstore?

Reviewer 4: I'm always hesitant when I see very prescriptive security requirements, as best practices change over time. What if we struck this line and add "... including secret scanning and code scanning" to "TI Follows security best practices..." above?

Author: Thanks a good suggestion, do you want to propose that wording change?

>  * Projects should be Securing Code Repository -> Managing Contributions Commit Signing , Secret Scanning, Code Scanning (OSFUZZ at a minimum) + Self-assessment Should OpenSSF require these if the SCM supports it, especially using Sigstore?

Reviewer 4: ```suggestion
```
Incorporated this into line 46 suggestion, above.

### Gets/Benefits

 TI eligible to receive all Gets from Sandbox plus:

>  * Receives infrastructure support

Reviewer 5: qq: is infrastructure support in the sense of cloud hosting, etc. I think there is a similar point in both sandbox and graduated, is the infrastructure support different? E.g. up to $X per project?

>  * Receives consideration as in-scope for any submission to an OpenSSF-managed conference or event.

Reviewer 5: Having a booth at a conference is something that seems very valuable for a project. is there consideration for this as well?

Author: Broadly, under Graduated we state "Receives consideration as in-scope for any submission to an OpenSSF-managed conference or event."  We can expand to be more explicit about booth space.

 * Project may request custom OpenSSF Logo for group
 * Projects may use the OpenSSF logo to promote their project (in accordance with the trademark guidelines). Projects may not be referred to as an "OpenSSF Project" or "OpenSSF $ProjectName." Projects may say they are an "OpenSSF Incubating Project."
 * With additional TAC or WG approval, may fundraise for dedicated project funds, coordinated by the OpenSSF.
 * Receives support with vulnerability disclosure from the OpenSSF (Vulnerability Disclosure WG).
 * May post project updates and tutorials to the OpenSSF blog.

## Graduated level Gives & Gets

> ### Gives/Requirements

Reviewer 5: I noticed that all the requirements are mainly around project governance and technical implementation, is user adoption also considered in the requirements for graduation?

 All requirements of Incubating must be fulfilled and additionally:
 * Projects must be able to show a consistent release cadence.
 * Maintains a point of contact for vulnerability reports and follow coordinated vulnerability disclosure practices.
 * Implements, practices, and refines mature software development and release practices, such as adherence to semantic versioning, and having a declared policy for stable releases and backported fixes.
 * Projects must have documented project governance and be able to demonstrate that governance in action.
 * When applicable, projects must have completed a security audit through a third party and addressed audit findings and recommendations.

>  *  When applicable, Projects should achieve **BLAH** level of SLSA

Reviewer 3: SLSA Level 3?

Reviewer 2: yeah, that's also what I had in mind but I am wondering whether we need to say SLSA Build Level 3 until more is defined.

Reviewer 4: Like I commented above, very specific security requirements tend not to age well. As it stands, there are multiple interpretations of "prevent secret material used to sign the provenance from being accessible to the user-defined build steps" which is part of SLSA v1.0 Build Level 3.

... but I don't have a great suggestion here. Maybe leave this off? Or say something like "Projects should harden their build systems in accordance with the SLSA Framework"?

>  *  When applicable, Projects should achieve **BLAH** level of SLSA

Reviewer 4: ```suggestion
 * Projects should harden their build systems in accordance with the SLSA Framework
```
